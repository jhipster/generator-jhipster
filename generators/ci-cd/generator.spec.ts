/**
 * Copyright 2013-2025 the original author or authors from the JHipster project.
 *
 * This file is part of the JHipster project, see https://www.jhipster.tech/
 * for more information.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { basename, dirname } from 'path';
import { fileURLToPath } from 'url';
import { describe, expect, it } from 'esmocha';
import { snakeCase } from 'lodash-es';

import { shouldSupportFeatures, testBlueprintSupport } from '../../test/support/tests.js';
import { defaultHelpers as helpers, runResult } from '../../lib/testing/helpers.js';
import Generator from './index.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const generator = basename(__dirname);

describe(`generator - ${generator}`, () => {
  it('generator-list constant matches folder name', async () => {
    await expect((await import('../generator-list.js'))[`GENERATOR_${snakeCase(generator).toUpperCase()}`]).toBe(generator);
  });
  shouldSupportFeatures(Generator);
  describe('blueprint support', () => testBlueprintSupport(generator));

  describe('cli', () => {
    describe('without ciCd values', () => {
      before(async () => {
        await helpers.runCli('ci-cd').withJHipsterConfig().withSkipWritingPriorities();
      });

      it('should match prompts order', () => {
        expect(runResult.askedQuestions.map(({ name }) => name)).toMatchInlineSnapshot(`
[
  "ciCd",
  "ciCdIntegrations",
]
`);
      });
    });

    describe('with invalid ciCd value', () => {
      it('should exit with error', async () => {
        await expect(
          helpers.runCli('ci-cd foo').withJHipsterConfig().withSkipWritingPriorities(),
        ).rejects.toThrowErrorMatchingInlineSnapshot(
          `"error: command-argument value 'foo' is invalid for argument 'ciCd'. Allowed choices are github, jenkins, gitlab, azure, travis, circle."`,
        );
      });
    });

    describe('with multiples values', () => {
      before(async () => {
        await helpers.runCli('ci-cd github jenkins gitlab azure').withJHipsterConfig().withSkipWritingPriorities();
      });

      it('should match context snapshot', () => {
        expect(runResult.generator.context).toMatchInlineSnapshot(`
{
  "artifactoryReleasesId": "releases",
  "artifactoryReleasesUrl": "http://artifactory:8081/artifactory/libs-release",
  "artifactorySnapshotsId": "snapshots",
  "artifactorySnapshotsUrl": "http://artifactory:8081/artifactory/libs-snapshot",
  "ciCd": [
    "github",
    "jenkins",
    "gitlab",
    "azure",
  ],
  "ciCdAny": true,
  "ciCdAzure": true,
  "ciCdCircle": false,
  "ciCdGithub": true,
  "ciCdGitlab": true,
  "ciCdIntegrations": [],
  "ciCdIntegrationsAny": false,
  "ciCdIntegrationsCypressDashboard": false,
  "ciCdIntegrationsDeploy": false,
  "ciCdIntegrationsHeroku": false,
  "ciCdIntegrationsPublishDocker": false,
  "ciCdIntegrationsSnyk": false,
  "ciCdIntegrationsSonar": false,
  "ciCdJenkins": true,
  "ciCdTravis": false,
  "dockerImage": undefined,
  "frontTestCommand": "test",
  "gitLabIndent": "",
  "herokuAppName": "jhipster",
  "indent": "",
  "insideDocker": false,
  "sendBuildToGitlab": false,
  "sonarName": "sonar",
  "sonarOrga": undefined,
  "sonarUrl": "https://sonarcloud.io",
}
`);
      });
      it('should populate context', () => {
        expect(runResult.generator.context!.ciCd).toEqual(['github', 'jenkins', 'gitlab', 'azure']);
      });
    });

    describe('with github', () => {
      before(async () => {
        await helpers.runCli('ci-cd github').withJHipsterConfig().withSkipWritingPriorities();
      });

      it('should not ask ciCd question', () => {
        expect(runResult.askedQuestions.map(({ name }) => name)).toMatchInlineSnapshot(`
[
  "ciCdIntegrations",
]
`);
      });
    });

    describe('with jenkins', () => {
      before(async () => {
        await helpers.runCli('ci-cd jenkins').withJHipsterConfig().withSkipWritingPriorities();
      });

      it('should not ask ciCd question', () => {
        expect(runResult.askedQuestions.map(({ name }) => name)).toMatchInlineSnapshot(`
[
  "ciCdIntegrations",
  "insideDocker",
  "sendBuildToGitlab",
]
`);
      });
    });
  });
});
