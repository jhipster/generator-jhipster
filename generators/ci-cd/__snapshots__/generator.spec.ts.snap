// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator - ci-cd cli with multiples values should match context snapshot 1`] = `
{
  "addOpenapiGeneratorPlugin": undefined,
  "additionalSubGenerators": undefined,
  "anyEntityHasRelationshipWithUser": false,
  "applicationType": "monolith",
  "applicationTypeAny": true,
  "applicationTypeGateway": false,
  "applicationTypeMicroservice": false,
  "applicationTypeMonolith": true,
  "artifactoryReleasesId": "releases",
  "artifactoryReleasesUrl": "http://artifactory:8081/artifactory/libs-release",
  "artifactorySnapshotsId": "snapshots",
  "artifactorySnapshotsUrl": "http://artifactory:8081/artifactory/libs-snapshot",
  "authenticationType": "jwt",
  "authenticationTypeAny": true,
  "authenticationTypeJwt": true,
  "authenticationTypeOauth2": false,
  "authenticationTypeSession": false,
  "authenticationUsesCsrf": false,
  "authority": Any<Object>,
  "autoCrlf": undefined,
  "backendType": "Java",
  "backendTypeJavaAny": true,
  "backendTypeSpringBoot": true,
  "baseName": "jhipster",
  "buildTool": "maven",
  "buildToolAny": true,
  "buildToolGradle": false,
  "buildToolMaven": true,
  "buildToolUnknown": false,
  "cacheProvider": "ehcache",
  "cacheProviderAny": true,
  "cacheProviderCaffeine": false,
  "cacheProviderEhcache": true,
  "cacheProviderHazelcast": false,
  "cacheProviderInfinispan": false,
  "cacheProviderMemcached": false,
  "cacheProviderNo": false,
  "cacheProviderRedis": false,
  "camelizedBaseName": "jhipster",
  "capitalizedBaseName": "Jhipster",
  "caret": undefined,
  "ciCd": [
    "github",
    "jenkins",
    "gitlab",
    "azure",
  ],
  "ciCdAny": true,
  "ciCdAzure": true,
  "ciCdCircle": false,
  "ciCdGithub": true,
  "ciCdGitlab": true,
  "ciCdIntegrations": [],
  "ciCdIntegrationsAny": false,
  "ciCdIntegrationsCypressDashboard": false,
  "ciCdIntegrationsDeploy": false,
  "ciCdIntegrationsHeroku": false,
  "ciCdIntegrationsPublishDocker": false,
  "ciCdIntegrationsSnyk": false,
  "ciCdIntegrationsSonar": false,
  "ciCdJenkins": true,
  "ciCdTravis": false,
  "cli": undefined,
  "cliName": undefined,
  "clientBundler": "webpack",
  "clientBundlerAny": true,
  "clientBundlerExperimentalEsbuild": false,
  "clientBundlerName": undefined,
  "clientBundlerVite": false,
  "clientBundlerWebpack": true,
  "clientDistDir": "target/classes/static/",
  "clientFramework": "angular",
  "clientFrameworkAngular": true,
  "clientFrameworkAny": true,
  "clientFrameworkBuiltIn": true,
  "clientFrameworkNo": false,
  "clientFrameworkReact": false,
  "clientFrameworkVue": false,
  "clientPackageJsonScripts": {},
  "clientPackageManager": "npm",
  "clientRootDir": "",
  "clientSrcDir": "src/main/webapp/",
  "clientTestDir": "src/test/javascript/",
  "clientTestFramework": undefined,
  "clientTestFrameworkName": undefined,
  "clientTestFrameworks": undefined,
  "clientTestFrameworksAny": false,
  "clientTestFrameworksCypress": false,
  "clientTheme": "none",
  "clientThemeVariant": "",
  "clientThemeVariantAny": false,
  "clientThemeVariantDark": false,
  "clientThemeVariantLight": false,
  "clientThemeVariantPrimary": false,
  "communicationSpringWebsocket": undefined,
  "creationTimestamp": 1577836800000,
  "cucumberTests": false,
  "customizeTemplatePaths": [],
  "cypressAudit": undefined,
  "cypressCoverage": undefined,
  "cypressTests": false,
  "dasherizedBaseName": "jhipster",
  "databaseMigration": "liquibase",
  "databaseMigrationAny": true,
  "databaseMigrationLiquibase": true,
  "databaseMigrationNo": false,
  "databaseType": "sql",
  "databaseTypeAny": true,
  "databaseTypeCassandra": false,
  "databaseTypeCouchbase": false,
  "databaseTypeMongodb": false,
  "databaseTypeNeo4j": false,
  "databaseTypeNo": false,
  "databaseTypeSql": true,
  "db": undefined,
  "defaultEnvironment": undefined,
  "defaultEnvironmentAny": false,
  "defaultEnvironmentDev": false,
  "defaultEnvironmentProd": false,
  "defaultPackaging": undefined,
  "defaultPackagingAny": false,
  "defaultPackagingJar": false,
  "defaultPackagingWar": false,
  "devDatabasePassword": undefined,
  "devDatabaseType": "postgresql",
  "devDatabaseTypeAny": true,
  "devDatabaseTypeH2Any": false,
  "devDatabaseTypeH2Disk": false,
  "devDatabaseTypeH2Memory": false,
  "devDatabaseTypeMariadb": false,
  "devDatabaseTypeMssql": false,
  "devDatabaseTypeMysql": false,
  "devDatabaseTypeOracle": false,
  "devDatabaseTypePostgresql": true,
  "devDatabaseUsername": undefined,
  "devJdbcUrl": undefined,
  "devServerPort": 4200,
  "devServerPortProxy": 9000,
  "dockerContainers": Any<Object>,
  "dockerImage": undefined,
  "dockerServicesDir": "src/main/docker/",
  "documentationArchiveUrl": "https://www.jhipster.tech/documentation-archive/vJHIPSTER_VERSION",
  "domains": [
    "com.mycompany.myapp",
  ],
  "dtoSuffix": "DTO",
  "dynamic": undefined,
  "embeddableLaunchScript": undefined,
  "enableGradleDevelocity": undefined,
  "enableHibernateCache": true,
  "enableI18nRTL": false,
  "enableSwaggerCodegen": false,
  "enableTranslation": true,
  "endpointPrefix": "",
  "entityPackages": [
    "com.mycompany.myapp",
  ],
  "entitySuffix": "",
  "feignClient": undefined,
  "filterEntitiesAndPropertiesForClient": [Function],
  "filterEntitiesForClient": [Function],
  "filterEntityPropertiesForClient": [Function],
  "frontendAppName": "jhipsterApp",
  "gatewayRoutes": undefined,
  "gatewayServerPort": undefined,
  "gatewayServicesApiAvailable": undefined,
  "gatlingTests": false,
  "generateAuthenticationApi": true,
  "generateBuiltInAuthorityEntity": true,
  "generateBuiltInUserEntity": true,
  "generateInMemoryUserCredentials": false,
  "generateUserManagement": true,
  "gitLabIndent": "",
  "githubRepository": undefined,
  "githubWorkflows": undefined,
  "graalvmSupport": undefined,
  "gradleDevelocityHost": undefined,
  "hasNonBuiltInEntity": false,
  "herokuAppName": "jhipster",
  "hipster": "jhipster_family_member_3",
  "hipsterBugTrackerLink": "https://github.com/jhipster/generator-jhipster/issues?state=open",
  "hipsterBugTrackerProductName": "JHipster",
  "hipsterChatLink": "https://gitter.im/jhipster/generator-jhipster",
  "hipsterChatProductName": "JHipster",
  "hipsterDocumentationLink": "https://www.jhipster.tech/",
  "hipsterHomePageProductName": "JHipster",
  "hipsterName": "Java Hipster",
  "hipsterProductName": "JHipster",
  "hipsterProjectLink": "https://github.com/jhipster/generator-jhipster",
  "hipsterStackOverflowProductName": "JHipster",
  "hipsterStackoverflowLink": "https://stackoverflow.com/tags/jhipster/info",
  "hipsterTwitterLink": "https://twitter.com/jhipster",
  "hipsterTwitterUsername": "@jhipster",
  "humanizedBaseName": "JHipster",
  "imperativeOrReactive": "imperative",
  "incrementalChangelog": undefined,
  "indent": "",
  "insideDocker": false,
  "javaCompatibleVersions": [
    "17",
    "21",
    "24",
  ],
  "javaDependencies": Any<Object>,
  "javaManagedProperties": {},
  "javaNodeBuildPaths": [],
  "javaPackageSrcDir": "src/main/java/com/mycompany/myapp/",
  "javaPackageTestDir": "src/test/java/com/mycompany/myapp/",
  "javaProperties": {},
  "javaVersion": "JAVA_VERSION",
  "jdlStore": undefined,
  "jhiPrefix": "jhi",
  "jhiPrefixCapitalized": "Jhi",
  "jhiPrefixDashed": "jhi",
  "jhiTablePrefix": "jhi",
  "jhipsterDependenciesVersion": "JHIPSTER_DEPENDENCIES_VERSION",
  "jhipsterPackageJson": Any<Object>,
  "jhipsterVersion": "JHIPSTER_VERSION",
  "js": undefined,
  "jsLoginRegex": "^[a-zA-Z0-9!$&*+=?^_\`{|}~.-]+@[a-zA-Z0-9-]+(?:\\\\.[a-zA-Z0-9-]+)*$|^[_.@A-Za-z0-9-]+$",
  "jwtSecretKey": undefined,
  "languages": [],
  "languagesDefinition": [],
  "listOrFlux": "List",
  "listOrFluxClassPath": "java.util.List",
  "localBlueprint": undefined,
  "loginRegex": "^(?>[a-zA-Z0-9!$&*+=?^_\`{|}~.-]+@[a-zA-Z0-9-]+(?:\\\\.[a-zA-Z0-9-]+)*)|(?>[_.@A-Za-z0-9-]+)$",
  "lowercaseBaseName": "jhipster",
  "mainClass": "JhipsterApp",
  "messageBroker": "no",
  "messageBrokerAny": false,
  "messageBrokerKafka": false,
  "messageBrokerNo": true,
  "messageBrokerPulsar": false,
  "microfrontend": false,
  "microfrontends": undefined,
  "monorepository": undefined,
  "nativeLanguage": "en",
  "nativeLanguageDefinition": {
    "angularLocale": "en",
    "dayjsLocale": "en",
    "displayName": "English",
    "fakerjsLocale": "en",
    "javaLocaleMessageSourceSuffix": "en",
    "languageTag": "en",
    "name": "English",
    "rtl": false,
  },
  "nodeDependencies": {
    "concurrently": "CONCURRENTLY_VERSION",
    "npm": "NPM_VERSION",
    "prettier": "PRETTIER_VERSION",
    "prettier-plugin-java": "PRETTIER_PLUGIN_JAVA_VERSION",
    "prettier-plugin-packagejson": "PRETTIER_PLUGIN_PACKAGEJSON_VERSION",
    "wait-on": "WAIT_ON_VERSION",
  },
  "nodePackageManager": "npm",
  "nodeVersion": "NODE_VERSION",
  "optionalOrMono": "Optional",
  "optionalOrMonoClassPath": "java.util.Optional",
  "optionalOrMonoOfNullable": "Optional.ofNullable",
  "packageFolder": "com/mycompany/myapp/",
  "packageInfoJavadocs": [
    {
      "documentation": "Logging aspect.",
      "packageName": "com.mycompany.myapp.aop.logging",
    },
    {
      "documentation": "Application management.",
      "packageName": "com.mycompany.myapp.management",
    },
    {
      "documentation": "Webflux database column mapper.",
      "packageName": "com.mycompany.myapp.repository.rowmapper",
    },
    {
      "documentation": "Application security utilities.",
      "packageName": "com.mycompany.myapp.security",
    },
    {
      "documentation": "Data transfer objects for rest mapping.",
      "packageName": "com.mycompany.myapp.service.dto",
    },
    {
      "documentation": "Data transfer objects mappers.",
      "packageName": "com.mycompany.myapp.service.mapper",
    },
    {
      "documentation": "Request chain filters.",
      "packageName": "com.mycompany.myapp.web.filter",
    },
    {
      "documentation": "Rest layer error handling.",
      "packageName": "com.mycompany.myapp.web.rest.errors",
    },
    {
      "documentation": "Rest layer visual models.",
      "packageName": "com.mycompany.myapp.web.rest.vm",
    },
  ],
  "packageJsonNodeEngine": undefined,
  "packageJsonScripts": {},
  "packageJsonType": undefined,
  "packageJsonTypeAny": false,
  "packageJsonTypeCommonjs": false,
  "packageJsonTypeModule": false,
  "packageName": "com.mycompany.myapp",
  "pkType": undefined,
  "prettierExtensions": "md,json,yml,html,cjs,mjs,js,ts,tsx,css,scss,java",
  "prettierTabWidth": undefined,
  "prodDatabasePassword": undefined,
  "prodDatabaseType": "postgresql",
  "prodDatabaseTypeAny": true,
  "prodDatabaseTypeMariadb": false,
  "prodDatabaseTypeMssql": false,
  "prodDatabaseTypeMysql": false,
  "prodDatabaseTypeOracle": false,
  "prodDatabaseTypePostgresql": true,
  "prodDatabaseUsername": undefined,
  "prodJdbcUrl": undefined,
  "projectDescription": "Description for JHipster",
  "projectVersion": "0.0.1-SNAPSHOT",
  "reactive": false,
  "rememberMeKey": undefined,
  "routes": undefined,
  "searchEngine": "no",
  "searchEngineAny": false,
  "searchEngineCouchbase": false,
  "searchEngineElasticsearch": false,
  "searchEngineNo": true,
  "sendBuildToGitlab": false,
  "serverPort": 8080,
  "serviceDiscovery": "no",
  "serviceDiscoveryAny": false,
  "serviceDiscoveryConsul": false,
  "serviceDiscoveryEureka": false,
  "serviceDiscoveryNo": true,
  "serviceDiscoveryType": "no",
  "serviceDiscoveryTypeAny": false,
  "serviceDiscoveryTypeConsul": false,
  "serviceDiscoveryTypeEureka": false,
  "serviceDiscoveryTypeNo": true,
  "skipCheckLengthOfIdentifier": undefined,
  "skipClient": undefined,
  "skipCommitHook": undefined,
  "skipFakeData": undefined,
  "skipJhipsterDependencies": undefined,
  "skipServer": undefined,
  "skipUserManagement": false,
  "sonarName": "sonar",
  "sonarOrga": undefined,
  "sonarUrl": "https://sonarcloud.io",
  "srcMainJava": "src/main/java/",
  "srcMainResources": "src/main/resources/",
  "srcMainWebapp": "src/main/webapp/",
  "srcTestJava": "src/test/java/",
  "srcTestJavascript": "src/test/javascript/",
  "srcTestResources": "src/test/resources/",
  "subGenerators": undefined,
  "syncUserWithIdp": undefined,
  "temporaryDir": "target/",
  "testFrameworks": [],
  "typescriptEslint": undefined,
  "upperFirstCamelCaseBaseName": "Jhipster",
  "useNpmWrapper": true,
  "user": Any<Object>,
  "userManagement": Any<Object>,
  "websocket": "no",
  "withAdminUi": true,
  "withGeneratedFlag": undefined,
  "wrapMono": [Function],
}
`;
